packages/grid/x-data-grid-pro/src/DataGridProVirtualScroller.tsx

```diff
@@ -204,6 +244,47 @@ const DataGridProVirtualScroller = React.forwardRef<
     minHeight: shouldExtendContent ? '100%' : 'auto',
   };
 
+  const { getDetailPanelContent } = rootProps;
+
+  const getDetailPanels = () => {
+    const panels: React.ReactNode[] = [];
+
+    if (getDetailPanelContent == null) {
+      return panels;
+    }
+
+    const rowsMeta = gridRowsMetaSelector(apiRef.current.state);
+    const uniqueExpandedRowIds = [...new Set([...expandedRowIds]).values()];
+
+    for (let i = 0; i < uniqueExpandedRowIds.length; i += 1) {
+      const id = uniqueExpandedRowIds[i];
+      const content = detailPanelsContent[id];
+
+      // Check if the id exists in the current page
+      const exists = currentPage.rows.find((row) => row.id === id);
```

Comment:
When pagination is disabled, this list can be very long so we should avoid looping through it twice.
We can store the `rowIndex` here instead of looking for is below, and do `if (React.isValidElement(content) && rowIndex > -1) {`

It would be even better to not have this lookup at all but I don't see an easy way of doing it.
An improvement if this seem to be a bottleneck would be to only loop the current page rows once for every panels
If the panel model was an object, it could look something like:

```tsx
const expandedPanelsRowIndexes: Record<GridRowId, number> = {}
for (let i=0; i<currentPage.rows.length; i+=1) {
  const id = currentPage.rows[i].id
  if (expandedPanelsLookup[id]) {
    expandedPanelsRowIndexes[id] = i
  }
}
```

And if the model remains an array, you can convert it to a lookup before.

That way even if you have 20 opened detail panels, you only loop on the rows once.

We could also avoid to re-run this method when the changing the rendered rows by putting the logic inside a `useMemo`

```ts
  const detailPanels = React.useMemo(() => {
    const panels: React.ReactNode[] = [];

    // Same logic

    return panels;
  }, [
    currentPage.rows,
    expandedRowIds,
    rowsMeta,
    classes.detailPanel,
    detailPanelsContent,
    detailPanelsHeights,
    getDetailPanelContent,
    rowHeight,
  ]);
```

Reply:
I created a lookup and memoized it. Memoizing `detailPanels` didn't work. If lets say `expandedRowIds` changes, it will be called again but `rowsMeta.positions` might have outdated values.